---
title: "COGS_IQ_pathway_DSM"
output: html_document
---

# SUMMARY for 15/03/17

* A new app has been created to view our results: SAVE-PRS: [**S**et **A**nalysis **V**iewing **E**nvironment - **P**olygenic **R**isk **S**cores](https://johnhubertjj.shinyapps.io/Viewing_PRS_two_files/)
* Allow comparisons across types of Polygenic risk scores, p-value thresholds, Categorical variables and 
* Step-wise regression has been calculated in for the DSM categories to test whether we can compare linear models across different DSM categories.
* A null vs alternative regression has also been calculated, but the results still need some formating
* Draft of secondary proposal has been attached to the email

* Full genome PRS has been calculated for CLOZUK_no_PGC2 and PGC2_no_CLOZUK to test sets against case/control phenotype
* Analysis on COGS re-done with IQ as a training set in each polygenic risk score.
* Results impending as we speak.

# RESULTS Summary

* Unfortunately, while both narrow and broad SCZ categorisations contributed to the differenecs in the polygenic risk score between samples, no significant associations were found around the optimum PRS P-value threshold (p = 0.05)
* But significant associations found for abnormal grooming behaviour and Cav2 channels (as well as the whole genome PRS of course)
* However, using a narrow DSM does appear consistently improve the whole genome PRS test against cognition but the models are quite poor. 

# Discussion points

* Run data-driven pathway analysis? (link up with Andrews annotations)
* Use a gene-PRS to decipher which segments are contributing to the signal in gene-set PRS?
* More narrow examination at this analysis?
* Secondary proposal for BD de-duplicated from SCZ
* Need to link up with Alex for larger cognition sample 
* Software help (c++?)

## INFORMATION

#### Polygenic risk score

**QC**

* SE > 5
* MAF < 0.1
* HWE 1e-04, mid-p, include nonctrl in COGS.
* genotype missingness thresh = 0.1
* Clumping
  * p1 = 1
  * p2 = 1
  * r2 = 0.1
  * window = 500kb
  
* Gene-set PRS have been limited to genic SNPs, then clumped within gene-set
* Whole genome PRS has been limited to genic SNPs
  
#### Sample sizes

* COGS = 1024
* IQ = **????**

#### Testing set Info
**?????**

* Sample info, NSNPs in the analysis  

#### Training set info  
**?????**

* Obtained from where? what year? what paper? who was involved?

#### Gene sets

* 5HT_2C
* abnormal behaviour
* abnormal grooming behaviour (such a weird set, you'll see why in a second)
* abnormal LTP
* abnormal nervous system electrophysiology
* Calcium ion import (GO)
* Cav2 channels
* FMRP targets
* LOF intolerance
* Membrane Depolarisation during action potential (GO)
* Synaptic transmission (GO)



* All taken from James + Antonio's paper apart from GO terms which were defined in Amigo GO
  * GO terms were part of a conditional analysis at the top of Antonio's/James' paper in a data-driven pathway analysis so I've included them here
  
#### Phenotypes

**DSM**

1. Schizophrenia
2. Schizoaffective depressed
3. Schizoaffective bipolar
4. Other psychotic disorder
5. Bipolar 1
6. Bipolar 2
7. Depression
8. Other (e.g. OCD)
9. Missing

* DSM 1-4 is classified as "broad" Schizophrenia
* DSM 1-2 is classified as "narrow" SCZ

**ICD**

+ Same format as DSM?
  
**Composite_Zimpute**

+ Composite measure from MATRICS. Imputed for missing data
  + MATRICS is the "IQ test for schizophrenia" and is mean zero centered
  + Composite of many metrics of various forms of cognition into one score
    
**FINAL_CLOZ_ever**

+ Has the individual ever received clozapine treatment?
  + Cleaned from interview and self-reported
  + If self-reported stated 'no' for in the interview but 'yes' to clozapine, the reported answer was 'yes'
    
1 Yes  
0 No  
9 Missing  

**FINAL_TRS**

* Does the individual have treatment-resistant schizophrenia?
  * Cleaned from OPCRIT 89 (response to neuroleptics) and FINAL_Cloz_Ever
  * Classed as TRS if 'yes' to either above 
    
1 TRS  
0 non-TRS  
9 Missing  
  
* Only a few missing from TRS compared to clozapine so both TRS measures will correlate highly if used

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Insert Libraries

```{r, message=FALSE, include=F, results = 'hide'}
# required packages -- will only install if they are not already installed
list.of.packages <- c("plyr", "stringr", "dplyr", "tidyr", "reshape2", "ggplot2", "scales", "data.table", "plotly", "devtools", "gridExtra", "cowplot", "repr", "knitr", "kableExtra", "IRdisplay")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)


#library(ggplot2) 
    
# loads the required packages
lapply(list.of.packages, require, character.only = TRUE)

library(data.table)
library(plyr)
library(dplyr)
library(rms)
```
# Write the Name of the analysis below 
```{r}

```


## Files I need

Files that are required to be read into this analysis:

* COGS duplicates with PGC
* COGS covariates
* Pathway results with COGS and CLOZUK_PGC2_noCOGS summarised
* Phenotype files with DSM information and Cognition and treatment resistance
* COGS gene boundaries (normal) gene set polygenic risk scores
* COGS regulatory regions (+ 10kb, -35kb from gene boundary; extended) gene set polygenic risk scores
* COGS whole genome genic (gene boundaries = normal; regulatory regions = extended) polygenic risk scores
* COGS whole genome polygenic risk scores

#### Read in whole genome normal and extended genic files before separating
```{r}

# Output arguments and significance threshold arguments
Name_of_analysis <- "CLOZUKminusPGC2_PGC2_minusCLOZUK_SCZ_Gene-sets_case_control"
significance_threshold <- c(5e-08,1e-06,1e-04,0.01,0.05,0.1,0.2,0.5,1)
Location_of_analysis <- ("~/Documents/CLOZUK_PGC2_pathway_analysis/")

# Location of whole genome files
Location_of_whole_genome_normal_gene_files <- paste0("/Users/johnhubert/Documents/CLOZUK_PGC2_pathway_analysis/CLOZUKminusPGC2_PGC2minusCLOZUK_whole_genome_PRS/Genic_PRS/PRS_scoring/PGC2minusCLOZUK_CLOZUKminusPGC2_normal_gene_regions_Clumped_whole_genome_final_significance_threshold_at_",significance_threshold,".profile")

Location_of_whole_genome_extended_gene_files <- paste0("/Users/johnhubert/Documents/CLOZUK_PGC2_pathway_analysis/CLOZUKminusPGC2_PGC2minusCLOZUK_whole_genome_PRS/Genic_PRS/PRS_scoring/PGC2minusCLOZUK_CLOZUKminusPGC2_extended_gene_regions_Clumped_whole_genome_final_significance_threshold_at_",significance_threshold,".profile")

Location_of_whole_genome_entire_genome_files <-  paste0("/Users/johnhubert/Documents/CLOZUK_PGC2_pathway_analysis/CLOZUKminusPGC2_PGC2minusCLOZUK_whole_genome_PRS/Full_genome/PRS_scoring/PGC2minusCLOZUK_CLOZUKminusPGC2_whole_genome_significance_threshold_at_",significance_threshold,".profile")

# Location of gene-set PRS files
Location_of_extended_gene_set_PRS_files <- "/Users/johnhubert/Documents/CLOZUK_PGC2_pathway_analysis/SCHIZ_gene-sets_analysis/FINAL_PATHWAY_RESULTS_PRS_PROFILESPGC2_no_CLOZUK_CLOZUK_no_PGC2_Selected_Pocklington_plus_GO_pathways_SCHIZ_extended.txt"
Location_of_gene_boundary_gene_set_PRS_files <- "/Users/johnhubert/Documents/CLOZUK_PGC2_pathway_analysis/SCHIZ_gene-sets_analysis/FINAL_PATHWAY_RESULTS_PRS_PROFILESPGC2_no_CLOZUK_CLOZUK_no_PGC2_Selected_Pocklington_plus_GO_pathways_SCHIZ_normal.txt"

# Location of Phenotype and covariate files 
Location_of_Phenotype_file <- "/Users/johnhubert/Documents/CLOZUK_PGC2_pathway_analysis/Stationary_data/CLOZUK.r7.GWAS_IDs.fam"
# Location_of_Alternative_ID_file <- "~/Documents/COGS_pathways_results/COGS_IDs_V2.txt"
Location_of_Covariates <- "/Users/johnhubert/Documents/CLOZUK_PGC2_pathway_analysis/Stationary_data/CLOZUK2.r7.select2PC.eigenvec.txt"
#Location_of_Duplicate_samples <-"~/Documents/COGS_pathways_results/COGSv2016.PGCrelatives.txt"
Location_of_age_and_gender_phenotypes <-"~/Documents/COGS_pathways_results/CLOZUK.r7.GWAS_IDs.fam"


my_data <- lapply(Location_of_whole_genome_normal_gene_files, fread, header=TRUE) 
names(my_data) <- str_replace(Location_of_whole_genome_normal_gene_files, pattern = ".profile", replacement = "")

# iterate through significance thresholds 
for (i in 1:length(significance_threshold)) {
  my_data[[i]] <- my_data[[i]][,c(1,2,6)]
  colnames(my_data[[i]]) <- c("FID", "IID", paste("SCORE_whole_genome_", significance_threshold[i], sep=""))
}

all_prs_normal <- join_all(my_data, by=c("FID", "IID"), type='left')

cols <- names(all_prs_normal)[3:ncol(all_prs_normal)]
setnames(all_prs_normal, old = c(3:ncol(all_prs_normal)), new = paste("normal.genic.genome", cols, sep = "_"))


my_data <- lapply(Location_of_whole_genome_extended_gene_files, fread, header=TRUE) 
names(my_data) <- str_replace(Location_of_whole_genome_extended_gene_files, pattern = ".profile", replacement = "")

# iterate through significance thresholds 
for (i in 1:length(significance_threshold)) {
  my_data[[i]] <- my_data[[i]][,c(1,2,6)]
  colnames(my_data[[i]]) <- c("FID", "IID", paste("SCORE_whole_genome_", significance_threshold[i], sep=""))
}

all_prs_extended <- join_all(my_data, by=c("FID", "IID"), type='left')

cols <- names(all_prs_extended)[3:ncol(all_prs_extended)]
setnames(all_prs_extended, old = c(3:ncol(all_prs_extended)), new = paste("extended.genic.genome", cols, sep = "_"))

```

#### Read in whole genome; entire genome
```{r}

my_data <- lapply(Location_of_whole_genome_entire_genome_files, fread, header=TRUE) 
names(my_data) <- str_replace(Location_of_whole_genome_entire_genome_files, pattern = ".profile", replacement = "")

# iterate through significance thresholds 
for (i in 1:length(significance_threshold)) {
  my_data[[i]] <- my_data[[i]][,c(1,2,6)]
  colnames(my_data[[i]]) <- c("FID", "IID", paste("SCORE_whole_genome_", significance_threshold[i], sep=""))
}

all_prs_whole_genome <- join_all(my_data, by=c("FID", "IID"), type='left')
cols <- names(all_prs_whole_genome)[3:ncol(all_prs_whole_genome)]
setnames(all_prs_whole_genome, old = c(3:ncol(all_prs_whole_genome)), new = paste("All.genome", cols, sep = "_"))

```
### Add column patterns to distinguish between gene boundaries and regulatory regions

```{r}
COGS_Pathway_results_normal = fread(Location_of_gene_boundary_gene_set_PRS_files)
cols <- names(COGS_Pathway_results_normal)[3:ncol(COGS_Pathway_results_normal)]
setnames(COGS_Pathway_results_normal, old = c(3:ncol(COGS_Pathway_results_normal)), new = paste("normal_geneset", cols, sep = "_"))

COGS_Pathway_results_extended = fread(Location_of_extended_gene_set_PRS_files)
cols <- names(COGS_Pathway_results_extended)[3:ncol(COGS_Pathway_results_extended)]
setnames(COGS_Pathway_results_extended, old = c(3:ncol(COGS_Pathway_results_extended)), new = paste("extended_geneset", cols, sep = "_"))

```


#### Read in all input files

```{r}
# State COGS, but actually CLOZUK...need to generalise this script
COGS_Phenotypes = fread(Location_of_Phenotype_file)
colnames(COGS_Phenotypes) <- c("FID","IID","PID","MID","Sex","PHENO")


COGS_list <- list(
COGS_covaraties = fread(Location_of_Covariates),
#COGS_duplicates = fread(Location_of_Duplicate_samples, header = F),
#COGS_Pathway_results = fread("~/Documents/COGS_pathways_results/FINAL_PATHWAY_RESULTS_PRS_PROFILESCLOZUKPGC2_NO_COGS_COGS_Selected_Pocklington_plus_GO_pathways_SCHIZ.txt"),
COGS_Pathway_results_normal = COGS_Pathway_results_normal,
COGS_Pathway_results_extended = COGS_Pathway_results_extended,
#COGS_age_gender = fread(Location_of_age_and_gender_phenotypes),
COGS_whole_genome_genic_normal = all_prs_normal,
COGS_whole_genome_genic_extended = all_prs_extended,
COGS_whole_genome = all_prs_whole_genome,
#COGS_Pathway_test_randomisation <- fread("~/Documents/COGS_pathways_results/FINAL_PATHWAY_RESULTS_PRS_PROFILESCLOZUKPGC2_NO_COGS_COGS_abnormal_behaviorrandomised_gene_sets.txt")
COGS_Phenotypes = COGS_Phenotypes
)


```

## Restrict ID's to ones without duplicates/covariates/phenotype information

```{r}
#setnames(COGS_list$COGS_duplicates, c("FID", "IID"))
COGS_total <- join_all(dfs=COGS_list,by = "IID",type = "left")
setkey(COGS_total,IID)
#COGS_total <- COGS_total[!COGS_list$COGS_duplicates$IID]
COGS_total_copy <- copy(COGS_total)
```

## Scale all Polygenic risk scores
* Used the first five covariates in COGS, limited down to individuals for which we have covariates for...
* Standardised the polygenic risk score afterwards

```{r}
# PRS_tables <- COGS_total[, .SD, .SDcols = (names(COGS_total) %like% "SCORE" | names(COGS_total) %like% "PC")]

PRS_scores <- grep(x = names(COGS_total), pattern = "SCORE")
PRS_names <- colnames(COGS_total[, .SD, .SDcols = PRS_scores])
PC_s <- grep(x = names(COGS_total), pattern = "PC\\d+")

  for (l in PRS_scores){
    model0 <- glm(COGS_total[[l]] ~ PC1+PC2+PC3+PC4+PC5+PC6+PC9+PC11+PC12+PC13+PC19, family = gaussian, data = COGS_total)
    m1<-mean(residuals(model0))
    sd1<-sd(residuals(model0))
    NORMSCORE <- (residuals(model0)-m1)/sd1
    COGS_total[,(l) := NORMSCORE]
  }

```

## Limit to DSM (narrow and broad) and run multiple linear regressions on cognition

* SCZ all cognition = 772 individuals after QC
* SCZ all TRS =  812 individuals after QC

* SCZ narrow = 678 individuals after QC
* SCZ broad = 827 individuals after QC

* SCZ all minus narrow cognition = 182 after removing missing values
* SCZ narrow cognition = 586 after removing missing values
* SCZ broad cognition = 725 after removing missing values

* SCZ all minus narrow TRS = 172 after removing missing values
* SCZ narrow TRS = 638 after removing missing values
* SCZ broad TRS = 773 after removing missing values

## Limiting to age and gender phenotypes

* Age at interview
* Gender
  * 1 = male
  * 0 = female

* No NA values in covariates are seen after QC above (some are seen in full dataset but are removed by coincidence)

## Validity of DSM categorisations

In order to gauge whether it is valid to separate out individuals into DSM groups step-wise and constrained models were performed.

This is due to the fact that as you limit down to different DSM groups, you are also reducing the power of each linear regression for both whole genome and gene-set polygenic risk scores so more noise is entering the linear regression.

#### Step-wise regression protocol

* A new binary variable for each DSM category was defined whereby 1 = individual is included in the category and 0 = individual was not included in the category.

* For each Polygenic risk score (both whole genome and gene-set PRS), the following step-wise linear regression was performed:
    * PRS ~ Age at interview + Gender + narrow_DSM + broad_DSM + All_minus_narrow_DSM
        * direction = "both"
        * scope = always include Age and Gender based on prior hypothesis that age and gender will affect common genetic risk
        
* If any, the predictors for each PRS model with the lowest AIC were recorded for interpretation

#### Constrained model protocol

* A new binary variable for each DSM category was defined whereby 1 = individual is included in the category and 0 = individual was not included in the category.

* Four linear regressions were performed for each whole genome and gene-set PRS with one null model and 3 alternative models:
    * Null = Cognition phenotype ~ PRS + Age + gender
    * Alt1 = Cognition phenotype ~ PRS + Age + gender + Narrow_DSM
    * Alt2 = Cognition phenotype ~ PRS + Age + gender + Broad_DSM
    * Alt3 = Cognition phenotype ~ PRS + Age + gender + All_minus_narrow_DSM
    
* All four p-values will be analysed to see if the model improves with the added covariate against the null model.



```{r, message=FALSE, include = FALSE}

# create a score list
score_list <- PRS_names

if (all(COGS_total$PHENO == 1 | COGS_total$PHENO == 0) == FALSE){
COGS_total[,PHENO := PHENO - 1]
}

l <- "Case_control"

# this example if for a linear regression (the phenotype of interest is a quantiative trait)
# is using a discrete phenotype, a logistic regression needs to be run, and the code altered from 'lm' to 'glm' including the argument of 'family = binomial'
# alterations for the calculation of R2 will also need to be made using the command highlighted above



obj <- data.frame(test=0, score=0, estimate=0, SE=0, zvalue=0, p=0, r.squared.Nagel=0, lower = 0, upper = 0)

for (i in score_list) {
    fit <- lrm(PHENO ~ COGS_total[[i]] , data = COGS_total,se.fit = T)
    fit2 <- glm(PHENO ~ COGS_total[[i]] , data = COGS_total, family = "binomial")
    tmp <- coef(summary(fit2))
    
    SE <- sqrt(diag(vcov(fit)))
    SE <- SE[2]
    Coef <- fit$coefficients[2]
    
    CI <- confint(fit2, level=0.95)
    CI <- CI[2,]
    # upper CI for beta
    CI_upper <- exp((Coef)+(1.96*SE))
    # lower CI for beta
    CI_lower <- exp((Coef)-(1.96*SE))
    tmp2 <- fit$stats
    Nagel_r2 <- tmp2[c("R2")]
    
    tmp3 <- c("case_control",i,tmp[2,], Nagel_r2, CI)
    obj <- rbind(obj, tmp3)
}
  results <- obj[which(obj$score %in% score_list),]
  assign(paste0("results_",l), results)


```
#### NOTES

* For logistic regression, confit in R might produce different CI's than manual CI's in R see here: [http://people.upei.ca/hstryhn/stryhn208.pdf]
  * and here: [https://stats.stackexchange.com/questions/5304/why-is-there-a-difference-between-manually-calculating-a-logistic-regression-95]


#### TABLES

* Need to do these sets of analysis and plot each one for broad, narrow, all and allminusnarrow
  * compare pathways against a whole genome genic with normal regions
  * compare pathways against a whole genome genic with extended regions
  * compare normal pathways against a whole genome without regions
  * compare extended pathways against a whole genome without regions
  
* Also make secondary proposal for de-duplication against the PGC for Bipolar cases and controls
* Will convert all of this to a Shiny app
*   
```{r echo=F,results='asis', message=F, warning=F, include=F}


Full_data <- results_
Full_data <- as.data.table(Full_data)
Full_data[, samples.i. := test]

time_of_analysis <- paste(format(Sys.time(), "%d-%m-%Y_at_%H.%M"))


write.table(Full_data,file = paste0(Location_of_analysis,Name_of_analysis,"_",time_of_analysis,".txt"), row.names = F, quote = F)
```


