---
title: "COGS_pathway_DSM"
output: html_document
---

## INFORMATION

#### Polygenic risk score

**QC**

* SE > 5
* MAF < 0.1
* HWE 1e-04, mid-p, include nonctrl in COGS.
* genotype missingness thresh = 0.1
* Clumping
  * p1 = 1
  * p2 = 1
  * r2 = 0.1
  * window = 500kb
  
* Gene-set PRS have been limited to genic SNPs, then clumped within gene-set
* Whole genome PRS has been limited to genic SNPs
  
#### Sample sizes

* COGS = 1024
* CLOZUK_PGC2_NO_COGS \~ 104,000

#### Gene sets

* 5HT_2C
* abnormal behaviour
* abnormal grooming behaviour (such a weird set, you'll see why in a second)
* abnormal LTP
* abnormal nervous system electrophysiology
* Calcium ion import (GO)
* Cav2 channels
* FMRP targets
* LOF intolerance
* Membrane Depolarisation during action potential
* Synaptic transmission

* All taken from James + Antonio's paper apart from GO terms which were defined in Amigo GO

#### Phenotypes

**DSM**

1. Schizophrenia
2. Schizoaffective depressed
3. Schizoaffective bipolar
4. Other psychotic disorder
5. Bipolar 1
6. Bipolar 2
7. Depression
8. Other (e.g. OCD)
9. Missing

* DSM 1-4 is classified as "broad" Schizophrenia
* DSM 1-2 is classified as "narrow" SCZ

**ICD**

+ Same format as DSM?
  
**Composite_Zimpute**

+ Composite measure from MATRICS. Imputed for missing data
  + MATRICS is the "IQ test for schizophrenia" and is mean zero centered
  + Composite of many metrics of various forms of cognition into one score
    
**FINAL_CLOZ_ever**

+ Has the individual ever received clozapine treatment?
  + Cleaned from interview and self-reported
  + If self-reported stated 'no' for in the interview but 'yes' to clozapine, the reported answer was 'yes'
    
1 Yes  
0 No  
9 Missing  

**FINAL_TRS**

* Does the individual have treatment-resistant schizophrenia?
  * Cleaned from OPCRIT 89 (response to neuroleptics) and FINAL_Cloz_Ever
  * Classed as TRS if 'yes' to either above 
    
1 TRS  
0 non-TRS  
9 Missing  
  
* Only a few missing from TRS compared to clozapine so both TRS measures will correlate highly if used

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Insert Libraries

```{r, message=FALSE, include=F, results = 'hide'}
# required packages -- will only install if they are not already installed
list.of.packages <- c("plyr", "stringr", "dplyr", "tidyr", "reshape2", "ggplot2", "scales", "data.table", "plotly", "devtools", "gridExtra", "cowplot", "repr", "knitr", "kableExtra", "IRdisplay")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)


#library(ggplot2) 
    
# loads the required packages
lapply(list.of.packages, require, character.only = TRUE)

library(data.table)
library(plyr)
library(dplyr)
library(rms)
```


## Files I need

Files that are required to be read into this analysis:

* COGS duplicates with PGC
* COGS covariates
* Pathway results with COGS and CLOZUK_PGC2_noCOGS summarised
* Phenotype files with DSM information and Cognition and treatment resistance

#### Read in whole genome genic files
```{r}
significance_threshold <- c(5e-08,1e-06,1e-04,0.01,0.05,0.1,0.2,0.5,1)

my_files <- paste0("~/Documents/COGS_pathways_results/PRS_scoring/CLOZUKPGC2_NO_COGS_COGS_extended_gene_regions_Clumped_whole_genome_final_significance_threshold_at_",significance_threshold, ".profile")


my_data <- lapply(my_files, fread, header=TRUE) 
names(my_data) <- str_replace(my_files, pattern = ".profile", replacement = "")

# iterate through significance thresholds 
for (i in 1:length(significance_threshold)) {
  my_data[[i]] <- my_data[[i]][,c(1,2,6)]
  colnames(my_data[[i]]) <- c("FID", "IID", paste("SCORE_whole_genome_", significance_threshold[i], sep=""))
}

all_prs <- join_all(my_data, by=c("FID", "IID"), type='left')

```
#### Read in all input files

```{r}

COGS_Phenotypes = fread("~/Documents/COGS_pathways_results/Phenotypes.txt")
COGS_ID_S = fread("~/Documents/COGS_pathways_results/COGS_IDs_V2.txt")
setnames(COGS_Phenotypes, old = "ID", new = "Study_ID")
COGS_Phenotypes <- merge(COGS_Phenotypes,COGS_ID_S, by = "Study_ID")

COGS_list <- list(
COGS_covaraties = fread("~/Documents/COGS_pathways_results/COGSv2016.noREL.eigenvec"),
COGS_duplicates = fread("~/Documents/COGS_pathways_results/COGSv2016.PGCrelatives[2] copy.txt", header = F),
COGS_Pathway_results = fread("~/Documents/COGS_pathways_results/FINAL_PATHWAY_RESULTS_PRS_PROFILESCLOZUKPGC2_NO_COGS_COGS_Selected_Pocklington_plus_GO_pathways_SCHIZ.txt"),
COGS_whole_genome_results = all_prs,
# COGS_Pathway_test_randomisation <- fread("~/Documents/COGS_pathways_results/FINAL_PATHWAY_RESULTS_PRS_PROFILESCLOZUKPGC2_NO_COGS_COGS_abnormal_behaviorrandomised_gene_sets.txt")
COGS_Phenotypes = COGS_Phenotypes
)


```

## Restrict ID's to ones without duplicates/covariates/phenotype information

```{r}
setnames(COGS_list$COGS_duplicates, c("FID", "IID"))
COGS_total <- join_all(dfs=COGS_list,by = "IID",type = "left")
setkey(COGS_total,IID)
COGS_total <- COGS_total[!COGS_list$COGS_duplicates$IID]
COGS_total_copy <- copy(COGS_total)
```

## Scale all Polygenic risk scores
* Used the first five covariates in COGS, limited down to individuals for which we have covariates for...
* Standardised the polygenic risk score afterwards

```{r}
# PRS_tables <- COGS_total[, .SD, .SDcols = (names(COGS_total) %like% "SCORE" | names(COGS_total) %like% "PC")]

PRS_scores <- grep(x = names(COGS_total), pattern = "SCORE")
PRS_names <- colnames(COGS_total[, .SD, .SDcols = PRS_scores])
PC_s <- grep(x = names(COGS_total), pattern = "PC\\d+")

  for (l in PRS_scores){
    model0 <- glm(COGS_total[[l]] ~ PC1 + PC2 + PC3 + PC4 + PC5, family = gaussian, data = COGS_total)
    m1<-mean(residuals(model0))
    sd1<-sd(residuals(model0))
    NORMSCORE <- (residuals(model0)-m1)/sd1
    COGS_total[,(l) := NORMSCORE]
  }

```

## Limit to DSM (narrow and broad) and run multiple linear regressions on cognition

* SCZ all cognition = 772 individuals after QC
* SCZ all TRS =  individuals after QC

* SCZ narrow = 678 individuals after QC
* SCZ broad = 827 individuals after QC

* SCZ narrow_cognition = 586 after removing missing values
* SCZ broad_cognition = 725 after removing missing values

* SCZ narrow_TRS = 638 after removing missing values
* SCZ broad_TRS = 773 after removing missing values


```{r, message=FALSE}
# create separted DF for narrow and broad SCZ

rows_to_keep_SCZ_narrow <- COGS_total[DSM_diagnosis_new == 1 | DSM_diagnosis_new == 2 ]
rows_to_keep_SCZ_broad <- COGS_total[DSM_diagnosis_new == 1 | DSM_diagnosis_new == 2 | DSM_diagnosis_new == 3 | DSM_diagnosis_new == 4]

COGS_cognition <- COGS_total[Composite_Zimpute != 999]
COGS_treatment_resistance <- COGS_total[FINAL_TRS != 9]

rows_to_keep_SCZ_narrow_cognition <- rows_to_keep_SCZ_narrow[Composite_Zimpute != 999]
rows_to_keep_SCZ_broad_cognition <- rows_to_keep_SCZ_broad[Composite_Zimpute != 999]
rows_to_keep_SCZ_narrow_TRS <- rows_to_keep_SCZ_narrow[FINAL_TRS != 9]
rows_to_keep_SCZ_broad_TRS <- rows_to_keep_SCZ_broad[FINAL_TRS != 9]

list_of_tests_cognition <- list(
  rows_to_keep_SCZ_narrow_cognition = as.data.frame(rows_to_keep_SCZ_narrow_cognition[,test:= Composite_Zimpute]),
  rows_to_keep_SCZ_broad_cognition = as.data.frame(rows_to_keep_SCZ_broad_cognition[,test := Composite_Zimpute]),
  COGS_cognition = as.data.frame(COGS_cognition[,test:= Composite_Zimpute])
  )  

list_of_tests_TRS <- list(
  rows_to_keep_SCZ_narrow_TRS = as.data.frame(rows_to_keep_SCZ_narrow_TRS[,test := FINAL_TRS]),
  rows_to_keep_SCZ_broad_TRS = as.data.frame(rows_to_keep_SCZ_broad_TRS[,test:= FINAL_TRS]),
  COGS_treatment_resistance = as.data.frame(COGS_treatment_resistance[,test:= FINAL_TRS])
)

# create a score list
score_list <- PRS_names
names_of_analyses <- names(list_of_tests_cognition)


# this example if for a linear regression (the phenotype of interest is a quantiative trait)
# is using a discrete phenotype, a logistic regression needs to be run, and the code altered from 'lm' to 'glm' including the argument of 'family = binomial'
# alterations for the calculation of R2 will also need to be made using the command highlighted above
for (l in names_of_analyses){
obj <- data.frame(test=0, score=0, estimate=0, SE=0, tvalue=0, p=0, r.squared=0, lower = 0, upper = 0)
for (i in score_list) {
    fit <- lm(test ~ list_of_tests_cognition[[l]][,i],  data=list_of_tests_cognition[[l]])
    tmp <- coef(summary(fit))
    tmp2 <- summary(fit)
    true_r2 <- tmp2$r.squared
    CI <- confint(fit, level=0.95)
    CI <- CI[2,]
    tmp3 <- c(l,i,tmp[2,], true_r2, CI)
    obj <- rbind(obj, tmp3)
}
  results <- obj[which(obj$score %in% score_list),]
  assign(paste0("results_",l), results)
}

names_of_analyses <- names(list_of_tests_TRS)

for (l in names_of_analyses){
obj <- data.frame(test=0, score=0, estimate=0, SE=0, zvalue=0, p=0, r.squared.Nagel=0, lower = 0, upper = 0)
for (i in score_list) {
    fit <- lrm(test ~ list_of_tests_TRS[[l]][,i], data = list_of_tests_TRS[[l]],se.fit = T)
    fit2 <- glm(test ~ list_of_tests_TRS[[l]][,i], data = list_of_tests_TRS[[l]], family = "binomial")
    tmp <- coef(summary(fit2))
    
    SE <- sqrt(diag(vcov(fit)))
    SE <- SE[2]
    Coef <- fit$coefficients[2]
    
    CI <- confint(fit2, level=0.95)
    CI <- CI[2,]
    # upper CI for beta
    CI_upper <- exp((Coef)+(1.96*SE))
    # lower CI for beta
    CI_lower <- exp((Coef)-(1.96*SE))
    tmp2 <- fit$stats
    Nagel_r2 <- tmp2[c("R2")]
    
    tmp3 <- c(l,i,tmp[2,], Nagel_r2, CI)
    obj <- rbind(obj, tmp3)
}
  results <- obj[which(obj$score %in% score_list),]
  assign(paste0("results_",l), results)
}
```
#### NOTES

* For logistic regression, confit in R might produce different CI's than manual CI's in R see here: [http://people.upei.ca/hstryhn/stryhn208.pdf]
  * and here: [https://stats.stackexchange.com/questions/5304/why-is-there-a-difference-between-manually-calculating-a-logistic-regression-95]

* CURRENTLY, the whole genome genic used an R2 for clumping of 0.01 instead of 0.1, but results change was minimal when re-done for pathways, will still re-run in any case...

#### TABLES

```{r echo=F,results='asis', message=F}

all_cogs_PRS_results <- list(
  Cognition_all_samples = results_COGS_cognition,
  Cognition_broad = results_rows_to_keep_SCZ_broad_cognition,
  Cognition_narrow = results_rows_to_keep_SCZ_narrow_cognition
  )

all_TRS_PRS_results <- list (
  TRS_all_samples <- results_COGS_treatment_resistance,
  TRS_broad <- results_rows_to_keep_SCZ_broad_TRS,
  TRS_narrow <- results_rows_to_keep_SCZ_narrow_TRS
)
library(knitr)

samples <- c("ALL", "broad", "narrow")

for (t in 1:3){
current_table <- as.data.table(all_cogs_PRS_results[[t]])
current_table <- current_table[order(current_table$p)]
print(kable(current_table[1:20], caption = paste0("Cognition tables across different DSM ranges for ",samples[t]," samples")) %>%
kable_styling())
}

for (t in 1:3){
kable(all_TRS_PRS_results[[t]][1:20,], caption = "TRS tables across different DSM ranges")
}
```

### Plotting
```{r echo = F,message = F, results=F}

gglist_pathways <- list()
gglist_beta_pathways <- list()
gglist_r2_dir_pathways <- list()

for (t in 1:3){
  current_table <- as.data.table(all_cogs_PRS_results[[t]])

# Pathways
str(current_table)

COGS_only_0.5 <- current_table[grepl("_0.5",current_table$score)] 
COGS_only_0.05 <- current_table[grepl("_0.05",current_table$score)] 

#COGS_only_0.05 <- COGS_only_0.05[!grepl("abnormal_grooming_behavior", COGS_only_0.05$score)]
#COGS_only_0.5 <- COGS_only_0.5[!grepl("abnormal_grooming_behavior", COGS_only_0.5$score)]

#COGS_only_0.05 <- COGS_only_0.05[!grepl("Lek2015_LoFintolerant_90", COGS_only_0.05$score)]
#COGS_only_0.5 <- COGS_only_0.5[!grepl("Lek2015_LoFintolerant_90", COGS_only_0.5$score)]

split_data_frame_0.05 <- split(COGS_only_0.05, by = "test")
split_data_frame_0.5 <- split(COGS_only_0.5, by = "test")

#setkey(split_data_frame_0.05,p)
#setkey(split_data_frame_0.5,p)

#Whole brain
#ALSPAC_all_brain_0.05 <- ALSPAC_full_brain_regions[grepl("_0.05",ALSPAC_full_brain_regions$score)]
#ALSPAC_all_brain_0.5 <- ALSPAC_full_brain_regions[grepl("_0.5",ALSPAC_full_brain_regions$score)] 



for (i in 1:length(split_data_frame_0.05)){
  
  
  # Obtaining the log p-values
  top_five_pathways <- split_data_frame_0.05[[i]][order(split_data_frame_0.05[[i]]$score)]
  top_five_pathways$P_threshold <- 0.05
  top_five_pathways$Type <- c(rep("Pathway",11),"Whole_genome")
  top_five_pathways_0.05 <- which(top_five_pathways$score == "SCORE_whole_genome_0.05")
  
  # By Factoring the p-values, you can get them to trend decreasingly without affecting the whole genome p-value
  top_five_pathways$p <- as.numeric(top_five_pathways$p)
  top_five_pathways$score <- factor(top_five_pathways$score, levels = top_five_pathways$score[c(top_five_pathways_0.05, order(top_five_pathways$score[-top_five_pathways_0.05]))])
  
  
  # Do the same for the 0.5 p-thresholds
  top_five_pathways2 <- split_data_frame_0.5[[i]][order(split_data_frame_0.5[[i]]$score)]
  top_five_pathways2$P_threshold <- 0.5
  top_five_pathways2$Type <- c(rep("Pathway",11),"Whole_genome")
  top_five_pathways_0.5 <- which(top_five_pathways2$score == "SCORE_whole_genome_0.5")
  
  # By Factoring the p-values, you can get them to trend decreasingly without affecting the whole genome p-value
  top_five_pathways2$p <- as.numeric(top_five_pathways2$p)
  top_five_pathways2$score <- factor(top_five_pathways2$score, levels = top_five_pathways2$score[c(top_five_pathways_0.5, order(top_five_pathways2$score[-top_five_pathways_0.5]))])
  
  
  top_five_pathways_all <- rbind(top_five_pathways,top_five_pathways2)
  top_five_pathways_all$p <- as.numeric(top_five_pathways_all$p)
  top_five_pathways_all$logp <- -log10(top_five_pathways_all$p)
  
  top_five_pathways_all$estimate <- as.numeric(top_five_pathways_all$estimate)
  top_five_pathways_all$upper <- as.numeric(top_five_pathways_all$upper)
  top_five_pathways_all$lower <- as.numeric(top_five_pathways_all$lower)
  top_five_pathways_all$SE <- as.numeric(top_five_pathways_all$SE)
  
  top_five_pathways_all$SE_higher <- top_five_pathways_all$estimate + top_five_pathways_all$SE
  top_five_pathways_all$SE_lower <- top_five_pathways_all$estimate - top_five_pathways_all$SE
  
  top_five_pathways_all$r2_dir <- 100 * (as.numeric(top_five_pathways_all$r.squared) *
                                           (sign(as.numeric(top_five_pathways_all$estimate))))
  
  top_five_pathways_all$p_value_text <- paste("p =", scientific(top_five_pathways_all$p, digits = 2), sep = " ")
  assign(paste0(top_five_pathways_all$test[1],"_top_five_pathways_all_thresholds"), top_five_pathways_all, envir = .GlobalEnv)
  
  p <- ggplot(top_five_pathways_all, aes(x=score, y=logp, fill = Type, group=P_threshold))
  
  p <- p +
    geom_point(aes(colour = Type))
  
  test1 <- levels(p$data$score)
  whole_genome_positions <- grep(x = test1,pattern = "SCORE_whole_genome_\\d.\\d+",perl = T)
  alterations <- test1
  alterations[-whole_genome_positions] <- str_replace(string = test1[-whole_genome_positions], pattern = "SCORE_(.*)_\\d.\\d+",replacement = "\\1")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "long_term_potentiation",replacement = "LTP")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "action_potential",replacement = "AP")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "depolarization",replacement = "DP")
  alterations[whole_genome_positions] <- rep(x = "Whole Genome PRS", 2)
  
  test2 <- vector(length = length(test1))
  names(alterations) <- test1
  
  p <- p + scale_x_discrete(labels= alterations)
  #p <- p + scale_y
  p <- p + facet_grid(. ~ P_threshold,scales = "free_x", space = "free_x") +
    theme(strip.text.x = element_text(size = 10))
  p <- p + geom_hline(aes(yintercept=1.30103), colour = "red", linetype= "solid", alpha = 0.25)
  p <- p + scale_fill_brewer(palette = "Paired")
  p <- p + theme(axis.text.x = element_text(angle = 90,size = 10,hjust = 1,vjust = 0.5))
  p <- p + ggtitle(top_five_pathways_all$test[1])
  #p <- p + theme(plot.title = element_text( face = "bold",hjust = 0.5,vjust = 0.5,angle = 90))
  p <- p + theme(plot.title = element_text( face = "bold",hjust = 0.5))
  p <- p + ylab(label = expression(-'log'[10]*'(p)'))
  p <- p + xlab(label = "Polygenic risk score")

  gglist_pathways[[t]] <- p
  
  p <- ggplot(top_five_pathways_all, aes(x=score, y=estimate, fill = Type, group=P_threshold))
  
  p <- p +
    geom_errorbar(aes(ymin = SE_higher, ymax = SE_lower), position = "dodge", width = 0.25) +
    geom_point(aes(colour = Type))
  
  test1 <- levels(p$data$score)
  whole_genome_positions <- grep(x = test1,pattern = "SCORE_whole_genome_\\d.\\d+",perl = T)
  alterations <- test1
  alterations[-whole_genome_positions] <- str_replace(string = test1[-whole_genome_positions], pattern = "SCORE_(.*)_\\d.\\d+",replacement = "\\1")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "long_term_potentiation",replacement = "LTP")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "action_potential",replacement = "AP")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "depolarization",replacement = "DP")
  alterations[whole_genome_positions] <- rep(x = "Whole Genome PRS", 2)
  test2 <- vector(length = length(test1))
  names(alterations) <- test1
  
  p <- p + scale_x_discrete(labels= alterations)
  p <- p + facet_grid(. ~ P_threshold, scales = "free_x", space = "free_x") +
    theme(strip.text.x = element_text(size = 10))
  p <- p + geom_hline(aes(yintercept=0), colour = "red", linetype= "solid", alpha = 0.25)
  p <- p + scale_fill_brewer(palette = "Paired")
  p <- p + theme(axis.text.x = element_text(angle = 90,size = 10,hjust = 1,vjust = 0.5))
  p <- p + ggtitle(top_five_pathways_all$test[1])
  p <- p + theme(plot.title = element_text( face = "bold",hjust = 0.5))
  p <- p + ylab(label = "BETA")
  p <- p + xlab(label = "Polygenic risk score")
  
  gglist_beta_pathways[[t]] <- p
  
  p <- ggplot(top_five_pathways_all, aes(x=score, y=r2_dir, fill = Type, group=P_threshold))
  p <- p +
    geom_bar(stat = "identity", aes(colour = Type), position = "dodge") +
    geom_text(data=subset(top_five_pathways_all, p < 0.05),
              aes(x=score,y=r2_dir,label=p_value_text, hjust=ifelse(sign(r2_dir)>0, 0, 0)), angle = 90, position = position_dodge(width = 1), size = 2.9)

 #Problem with labels with a workaround
  # I use the score column in the format of factors and reference each relevant dataset for ggplot.
  # However this relies on having 0.05 and 0.5 in the value name.
  # scale_x_discrete accepts functions, but I also need to convert SCORE_0.05 and Score_0.5 into a "Whole_genome_PRS" which is almost impossible to write"
  # However as the labels function accepts key:value pairs, I wrote a vector in R that maps the original names of the pathways to "human readable" format using names function in R
  # This should work for most instances
  
  test1 <- levels(p$data$score)
  whole_genome_positions <- grep(x = test1,pattern = "SCORE_whole_genome_\\d.\\d+",perl = T)
  alterations <- test1
  alterations[-whole_genome_positions] <- str_replace(string = test1[-whole_genome_positions], pattern = "SCORE_(.*)_\\d.\\d+",replacement = "\\1")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "long_term_potentiation",replacement = "LTP")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "action_potential",replacement = "AP")
  alterations[-whole_genome_positions] <- str_replace(string = alterations[-whole_genome_positions], pattern = "depolarization",replacement = "DP")
  alterations[whole_genome_positions] <- rep(x = "Whole Genome PRS", 2)
  test2 <- vector(length = length(test1))
  names(alterations) <- test1
  
  p <- p + scale_x_discrete(labels= alterations)

  p <- p + scale_y_continuous(expand = expand_scale(mult = c(0.2,.6)))
  p <- p + facet_grid(. ~ P_threshold, scales = "free_x", space = "free_x") +
    theme(strip.text.x = element_text(size = 10))
  p <- p + theme(axis.text.x = element_text(angle = 90, size = 10, hjust = 1,vjust = 0.5))
  p <- p + ggtitle(top_five_pathways_all$test[1])
  p <- p + theme(plot.title = element_text( face = "bold",hjust = 0.5))
  p <- p + ylab(label = "R2_dir (%)")
  p <- p + xlab(label = "Polygenic risk score")
  
  gglist_r2_dir_pathways[[t]] <- p
  
  
}
}

gglist_beta_pathways
gglist_r2_dir_pathways

pdf("~/Documents/COGS_pathways_results/COGS_CLOZUK_pathway_whole_genome_genic_comparison_R2_dir_12_pathways.pdf", onefile = TRUE)
invisible(lapply(gglist_r2_dir_pathways,print))
dev.off()

pdf("~/Documents/COGS_pathways_results/COGS_CLOZUK_pathway_whole_genome_genic_comparison_beta_12_pathways.pdf", onefile = TRUE)
invisible(lapply(gglist_beta_pathways,print))
dev.off()


```


